From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: haotienh <haotienh@nvidia.com>
Date: Thu, 8 Aug 2024 08:31:34 +0000
Subject: [PATCH 148/240] NVIDIA: SAUCE: usb: gadget: xudc: Add platform init
 functions

Offsets and expected values of some registers for newer platforms are
different from previous ones.
Set registers in platform initializing functions for different
platforms.

Bug 4709389

Change-Id: I43e72af496b16305955f1490effafc10ad6ced25
Signed-off-by: haotienh <haotienh@nvidia.com>
Reviewed-on: https://git-master.nvidia.com/r/c/3rdparty/canonical/linux-noble/+/3248641
GVS: buildbot_gerritrpt <buildbot_gerritrpt@nvidia.com>
Reviewed-by: svcacv <svcacv@nvidia.com>
Reviewed-by: WK Tsai <wtsai@nvidia.com>
Reviewed-by: Henry Lin <henryl@nvidia.com>
Reviewed-by: EJ Hsu <ejh@nvidia.com>
(cherry picked from commit 265621324cc30874504d60870ae13ad4ed9e9a38)

diff --git a/drivers/usb/gadget/udc/tegra-xudc.c b/drivers/usb/gadget/udc/tegra-xudc.c
index 01567a9f2eb93..3ed44e8cc7129 100644
--- a/drivers/usb/gadget/udc/tegra-xudc.c
+++ b/drivers/usb/gadget/udc/tegra-xudc.c
@@ -150,7 +150,6 @@
 #define  SSPX_CORE_CNT30_LMPITP_TIMER_MASK GENMASK(19, 0)
 #define  SSPX_CORE_CNT30_LMPITP_TIMER(x) ((x) & \
 					SSPX_CORE_CNT30_LMPITP_TIMER_MASK)
-#define SSPX_CORE_CNT32 0x690
 #define  SSPX_CORE_CNT32_POLL_TBURST_MAX_MASK GENMASK(7, 0)
 #define  SSPX_CORE_CNT32_POLL_TBURST_MAX(x) ((x) & \
 					SSPX_CORE_CNT32_POLL_TBURST_MAX_MASK)
@@ -562,7 +561,7 @@ struct tegra_xudc_soc {
 	bool has_ipfs;
 	bool has_pg_support;
 	bool is_hv_guest;
-	bool skip_sspx_config;
+	void (*xudc_device_params_init)(struct tegra_xudc *xudc);
 };
 
 static inline u32 fpci_readl(struct tegra_xudc *xudc, unsigned int offset)
@@ -3427,47 +3426,10 @@ static void tegra_xudc_fpci_ipfs_init(struct tegra_xudc *xudc)
 	}
 }
 
-static void tegra_xudc_device_params_init(struct tegra_xudc *xudc)
+static void tegra_xudc_device_params_init_common(struct tegra_xudc *xudc)
 {
 	u32 val, imod;
 
-	val = xudc_readl(xudc, BLCG) & ~BLCG_COREPLL_PWRDN;
-	xudc_writel(xudc, val, BLCG);
-
-	if (xudc->soc->has_ipfs) {
-		val = xudc_readl(xudc, BLCG);
-		val |= BLCG_ALL;
-		val &= ~(BLCG_DFPCI | BLCG_UFPCI | BLCG_FE |
-				BLCG_COREPLL_PWRDN);
-		val |= BLCG_IOPLL_0_PWRDN;
-		val |= BLCG_IOPLL_1_PWRDN;
-		val |= BLCG_IOPLL_2_PWRDN;
-
-		xudc_writel(xudc, val, BLCG);
-	}
-
-	if (xudc->soc->port_speed_quirk)
-		tegra_xudc_limit_port_speed(xudc);
-
-	if (!xudc->soc->skip_sspx_config) {
-		/* Set a reasonable U3 exit timer value. */
-		val = xudc_readl(xudc, SSPX_CORE_PADCTL4);
-		val &= ~(SSPX_CORE_PADCTL4_RXDAT_VLD_TIMEOUT_U3_MASK);
-		val |= SSPX_CORE_PADCTL4_RXDAT_VLD_TIMEOUT_U3(0x5dc0);
-		xudc_writel(xudc, val, SSPX_CORE_PADCTL4);
-
-		/* Default ping LFPS tBurst is too large. */
-		val = xudc_readl(xudc, SSPX_CORE_CNT0);
-		val &= ~(SSPX_CORE_CNT0_PING_TBURST_MASK);
-		val |= SSPX_CORE_CNT0_PING_TBURST(0xa);
-		xudc_writel(xudc, val, SSPX_CORE_CNT0);
-
-		/* Default tPortConfiguration timeout is too small. */
-		val = xudc_readl(xudc, SSPX_CORE_CNT30);
-		val &= ~(SSPX_CORE_CNT30_LMPITP_TIMER_MASK);
-		val |= SSPX_CORE_CNT30_LMPITP_TIMER(0x978);
-		xudc_writel(xudc, val, SSPX_CORE_CNT30);
-	}
 	if (xudc->soc->lpm_enable) {
 		/* Set L1 resume duration to 95 us. */
 		val = xudc_readl(xudc, HSFSPI_COUNT13);
@@ -3476,17 +3438,6 @@ static void tegra_xudc_device_params_init(struct tegra_xudc *xudc)
 		xudc_writel(xudc, val, HSFSPI_COUNT13);
 	}
 
-	if (!xudc->soc->skip_sspx_config) {
-		/*
-		 * Compliance suite appears to be violating polling
-		 * LFPS tBurst max of 1.4us.  Send 1.45us instead.
-		 */
-		val = xudc_readl(xudc, SSPX_CORE_CNT32);
-		val &= ~(SSPX_CORE_CNT32_POLL_TBURST_MAX_MASK);
-		val |= SSPX_CORE_CNT32_POLL_TBURST_MAX(0xb0);
-		xudc_writel(xudc, val, SSPX_CORE_CNT32);
-	}
-
 	/* Direct HS/FS port instance to RxDetect. */
 	val = xudc_readl(xudc, CFG_DEV_FE);
 	val &= ~(CFG_DEV_FE_PORTREGSEL_MASK);
@@ -3536,6 +3487,81 @@ static void tegra_xudc_device_params_init(struct tegra_xudc *xudc)
 	xudc_writel(xudc, val, CFG_DEV_SSPI_XFER);
 }
 
+static void tegra186_xudc_device_params_init(struct tegra_xudc *xudc)
+{
+	u32 val;
+
+	val = xudc_readl(xudc, BLCG) & ~BLCG_COREPLL_PWRDN;
+	xudc_writel(xudc, val, BLCG);
+
+	if (xudc->soc->has_ipfs) {
+		val = xudc_readl(xudc, BLCG);
+		val |= BLCG_ALL;
+		val &= ~(BLCG_DFPCI | BLCG_UFPCI | BLCG_FE |
+				BLCG_COREPLL_PWRDN);
+		val |= BLCG_IOPLL_0_PWRDN;
+		val |= BLCG_IOPLL_1_PWRDN;
+		val |= BLCG_IOPLL_2_PWRDN;
+
+		xudc_writel(xudc, val, BLCG);
+	}
+
+	if (xudc->soc->port_speed_quirk)
+		tegra_xudc_limit_port_speed(xudc);
+
+	/* Set a reasonable U3 exit timer value. */
+	val = xudc_readl(xudc, SSPX_CORE_PADCTL4);
+	val &= ~(SSPX_CORE_PADCTL4_RXDAT_VLD_TIMEOUT_U3_MASK);
+	val |= SSPX_CORE_PADCTL4_RXDAT_VLD_TIMEOUT_U3(0x5dc0);
+	xudc_writel(xudc, val, SSPX_CORE_PADCTL4);
+
+	/* Default ping LFPS tBurst is too large. */
+	val = xudc_readl(xudc, SSPX_CORE_CNT0);
+	val &= ~(SSPX_CORE_CNT0_PING_TBURST_MASK);
+	val |= SSPX_CORE_CNT0_PING_TBURST(0xa);
+	xudc_writel(xudc, val, SSPX_CORE_CNT0);
+
+	/* Default tPortConfiguration timeout is too small. */
+	val = xudc_readl(xudc, SSPX_CORE_CNT30);
+	val &= ~(SSPX_CORE_CNT30_LMPITP_TIMER_MASK);
+	val |= SSPX_CORE_CNT30_LMPITP_TIMER(0x978);
+	xudc_writel(xudc, val, SSPX_CORE_CNT30);
+
+	/*
+	 * Compliance suite appears to be violating polling
+	 * LFPS tBurst max of 1.4us.  Send 1.45us instead.
+	 */
+#define SSPX_CORE_CNT32 0x690
+	val = xudc_readl(xudc, SSPX_CORE_CNT32);
+	val &= ~(SSPX_CORE_CNT32_POLL_TBURST_MAX_MASK);
+	val |= SSPX_CORE_CNT32_POLL_TBURST_MAX(0xb0);
+	xudc_writel(xudc, val, SSPX_CORE_CNT32);
+#undef SSPX_CORE_CNT32
+
+	tegra_xudc_device_params_init_common(xudc);
+}
+
+static void tegra264_xudc_device_params_init(struct tegra_xudc *xudc)
+{
+	u32 val;
+
+	val = xudc_readl(xudc, BLCG) & ~BLCG_COREPLL_PWRDN;
+	xudc_writel(xudc, val, BLCG);
+
+#define SSPX_CORE_CNT32 0x4090
+	/*
+	 * Compliance suite appears to be violating polling
+	 * LFPS tBurst max of 1.4us.  Send 1.466us instead.
+	 */
+	val = xudc_readl(xudc, SSPX_CORE_CNT32);
+	val &= ~(SSPX_CORE_CNT32_POLL_TBURST_MAX_MASK);
+	val |= SSPX_CORE_CNT32_POLL_TBURST_MAX(0x3a);
+	xudc_writel(xudc, val, SSPX_CORE_CNT32);
+#undef SSPX_CORE_CNT32
+
+	tegra_xudc_device_params_init_common(xudc);
+}
+
 static int tegra_xudc_phy_get(struct tegra_xudc *xudc)
 {
 	int err = 0, usb3_companion_port;
@@ -3694,6 +3720,7 @@ static struct tegra_xudc_soc tegra210_xudc_soc_data = {
 	.port_speed_quirk = false,
 	.has_ipfs = true,
 	.has_pg_support = true,
+	.xudc_device_params_init = tegra186_xudc_device_params_init,
 };
 
 static struct tegra_xudc_soc tegra186_xudc_soc_data = {
@@ -3709,6 +3736,7 @@ static struct tegra_xudc_soc tegra186_xudc_soc_data = {
 	.port_speed_quirk = false,
 	.has_ipfs = false,
 	.has_pg_support = true,
+	.xudc_device_params_init = tegra186_xudc_device_params_init,
 };
 
 static struct tegra_xudc_soc tegra194_xudc_soc_data = {
@@ -3724,6 +3752,7 @@ static struct tegra_xudc_soc tegra194_xudc_soc_data = {
 	.port_speed_quirk = true,
 	.has_ipfs = false,
 	.has_pg_support = true,
+	.xudc_device_params_init = tegra186_xudc_device_params_init,
 };
 
 static struct tegra_xudc_soc tegra234_xudc_soc_data = {
@@ -3738,6 +3767,7 @@ static struct tegra_xudc_soc tegra234_xudc_soc_data = {
 	.port_reset_quirk = false,
 	.has_ipfs = false,
 	.has_pg_support = true,
+	.xudc_device_params_init = tegra186_xudc_device_params_init,
 };
 
 static struct tegra_xudc_soc tegra234_hv_xudc_soc_data = {
@@ -3751,6 +3781,7 @@ static struct tegra_xudc_soc tegra234_hv_xudc_soc_data = {
 	.has_ipfs = false,
 	.has_pg_support = true,
 	.is_hv_guest = true,
+	.xudc_device_params_init = tegra186_xudc_device_params_init,
 };
 
 static struct tegra_xudc_soc tegra264_xudc_soc_data = {
@@ -3765,7 +3796,7 @@ static struct tegra_xudc_soc tegra264_xudc_soc_data = {
 	.port_reset_quirk = false,
 	.has_ipfs = false,
 	.has_pg_support = true,
-	.skip_sspx_config = true,
+	.xudc_device_params_init = tegra264_xudc_device_params_init,
 };
 
 static const struct of_device_id tegra_xudc_of_match[] = {
@@ -4133,7 +4164,7 @@ tegra_xudc_lowpower_exit(struct tegra_xudc *xudc)
 
 	tegra_xudc_fpci_ipfs_init(xudc);
 
-	tegra_xudc_device_params_init(xudc);
+	xudc->soc->xudc_device_params_init(xudc);
 
 	tegra_xudc_init_event_ring(xudc);
 
-- 
2.43.0

